Peer dependency error for `typescript` when initializing a new package

If you find a security vulnerability, do NOT open an issue. Email [security@sanity.io](mailto:security@sanity.io) instead.

**Describe the bug**

When using `npm` to initialize a new package:
```sh
npx @sanity/pkg-utils@latest init <pkg-name>
```

The installation will throw a peer dependency error:
```sh
npm ERR! code ERESOLVE
npm ERR! ERESOLVE unable to resolve dependency tree
npm ERR! 
npm ERR! While resolving: test@0.0.0
npm ERR! Found: typescript@5.4.3
npm ERR! node_modules/typescript
npm ERR!   dev typescript@"^5.4.3" from the root project
npm ERR! 
npm ERR! Could not resolve dependency:
npm ERR! peer typescript@"5.4.2" from @sanity/pkg-utils@5.1.7
npm ERR! node_modules/@sanity/pkg-utils
npm ERR!   dev @sanity/pkg-utils@"^5.1.7" from the root project
npm ERR! 
npm ERR! Fix the upstream dependency conflict, or retry
npm ERR! this command with --force or --legacy-peer-deps
npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
```

**Which versions of Node.js / npm are you running?**

Node v20.12.0
NPM v10.5.0

**Additional context**

`@sanity/pkg-utils` specifies an exact version of `typescript` as a peer dependency:
https://github.com/sanity-io/pkg-utils/blob/34952637ec612cb7f1939da92193c7dcb7c81252/package.json#L156

--- a/package.json
+++ b/package.json
@@ -129,6 +129,7 @@
     "@types/parse-git-config": "^3.0.4",
     "@types/prompts": "^2.4.9",
     "@types/rimraf": "^3.0.2",
+    "@types/semver": "^7.5.8",
     "@types/treeify": "^1.0.3",
     "@types/uuid": "^9.0.8",
     "@typescript-eslint/eslint-plugin": "^7.6.0",
@@ -148,13 +149,14 @@
     "outdent": "^0.8.0",
     "recast": "^0.23.6",
     "rollup-plugin-visualizer": "^5.12.0",
+    "semver": "^7.6.0",
     "ts-node": "^10.9.2",
     "typescript": "5.4.2",
     "vitest": "^1.4.0",
     "vitest-github-actions-reporter": "^0.11.1"
   },
   "peerDependencies": {
-    "typescript": "5.4.2"
+    "typescript": "5.4.x"
   },
   "packageManager": "pnpm@8.15.6",
   "engines": {

--- a/test/verifyTypeScriptVersion.test.ts
+++ b/test/verifyTypeScriptVersion.test.ts
@@ -1,19 +1,10 @@
 import apiExtractorJson from '@microsoft/api-extractor/package.json' assert {type: 'json'}
 import pkgUtilsJson from '@sanity/pkg-utils/package.json' assert {type: 'json'}
+import {satisfies} from 'semver'
 import {expect, test} from 'vitest'
 
-/**
- * Why are we requiring the same version of TypeScript as @microsoft/api-extractor?
- * @microsoft/api-extractor is shipping with `typescript` in its `dependencies`.
- * We have it as a peer dependency in @sanity/pkg-utils.
- * As long as everything results in the same version of `typescript`, we're good.
- * But if they're different then weird issues can happen, that can be incredibly hard to debug in userland.
- * To avoid all this we're requiring the same version of `typescript` as @microsoft/api-extractor at all times,
- * while it's still possible to override it, as adhering to `peerDependencies` largely depends on which package
- * manager is being used and how it's configured, you'll still at least get helpful warnings if there's a mismatch.
- */
-
-test('verify that @sanity/pkg-utils requires the same typescript version as @microsoft/api-extractor', () => {
-  expect(pkgUtilsJson.devDependencies.typescript).toBe(apiExtractorJson.dependencies.typescript)
-  expect(pkgUtilsJson.peerDependencies.typescript).toBe(apiExtractorJson.dependencies.typescript)
+test('verify that @sanity/pkg-utils requires the same typescript minor as @microsoft/api-extractor', () => {
+  expect(
+    satisfies(apiExtractorJson.dependencies.typescript, pkgUtilsJson.peerDependencies.typescript),
+  ).toBe(true)
 })

